File: packages/fastapi/src/grantex_fastapi/_middleware.py
"""FastAPI dependency-injection helpers for Grantex grant token verification."""

from typing import Any, Callable, Optional

from fastapi import Depends, Request
from fastapi.responses import JSONResponse
from grantex import GrantexTokenError, VerifiedGrant, VerifyGrantTokenOptions, verify_grant_token

from ._errors import ErrorCode, GrantexFastAPIError


def _extract_bearer_token(request: Request) -> Optional[str]:
    """Extract a Bearer token from the Authorization header."""
    header = request.headers.get("authorization")
    if not header:
        return None
    parts = header.split(" ", 1)
    if len(parts) != 2 or parts[0] != "Bearer":
        return None
    return parts[1]


class GrantexAuth:
    """Factory that produces FastAPI dependencies for Grantex token verification.

    Usage::

        from grantex_fastapi import GrantexAuth

        grantex = GrantexAuth(
            jwks_uri="https://grantex-auth-dd4mtrt2gq-uc.a.run.app/.well-known/jwks.json",
        )

        @app.get("/api/calendar")
        async def calendar(grant: VerifiedGrant = Depends(grantex)):
            return {"principalId": grant.principal_id, "scopes": grant.scopes}

    The instance is callable, so you can use ``Depends(grantex)`` directly.
    It can also produce scope-checking sub-dependencies via
    ``Depends(grantex.scopes("calendar:read"))``.
    """

    def __init__(
        self,
        jwks_uri: str,
        *,
        clock_tolerance: int = 0,
        audience: Optional[str] = None,
        token_extractor: Optional[Callable[[Request], Optional[str]]] = None,
    ) -> None:
        self._jwks_uri = jwks_uri
        self._clock_tolerance = clock_tolerance
        self._audience = audience
        self._token_extractor = token_extractor

    async def __call__(self, request: Request) -> VerifiedGrant:
        """FastAPI dependency that verifies the grant token and returns a VerifiedGrant."""
        return self._verify(request)

    def _verify(self, request: Request) -> VerifiedGrant:
        """Core verification logic shared by __call__ and scopes()."""
        if self._token_extractor is not None:
            token = self._token_extractor(request)
        else:
            token = _extract_bearer_token(request)

        if not token:
            raise GrantexFastAPIError(
                "TOKEN_MISSING",
                'No grant token found. Send a Grantex JWT in the Authorization header as "Bearer <token>".',
                401,
            )

        opts = VerifyGrantTokenOptions(
            jwks_uri=self._jwks_uri,
            clock_tolerance=self._clock_tolerance,
            audience=self._audience,
        )

        try:
            return verify_grant_token(token, opts)
        except GrantexTokenError as exc:
            msg = str(exc)
            is_expired = "exp" in msg.lower()
            code: ErrorCode = "TOKEN_EXPIRED" if is_expired else "TOKEN_INVALID"
            raise GrantexFastAPIError(code, msg, 401) from exc

    def scopes(self, *required_scopes: str) -> Callable[..., Any]:
        """Return a dependency that verifies the token AND checks scopes.

        Usage::

            @app.get("/api/calendar")
            async def calendar(grant: VerifiedGrant = Depends(grantex.scopes("calendar:read"))):
                ...

            # Multiple scopes — all required
            @app.post("/api/email/send")
            async def send(grant: VerifiedGrant = Depends(grantex.scopes("email:read", "email:send"))):
                ...
        """
        parent = self

        async def _dependency(request: Request) -> VerifiedGrant:
            grant = parent._verify(request)
            missing = [s for s in required_scopes if s not in grant.scopes]
            if missing:
                raise GrantexFastAPIError(
                    "SCOPE_INSUFFICIENT",
                    f"Grant token is missing required scopes: {', '.join(missing)}",
                    403,
                )
            return grant

        return _dependency


def require_scopes(grant: VerifiedGrant, *scopes: str) -> None:
    """Standalone scope check — call inside a route handler.

    Usage::

        @app.get("/api/data")
        async def data(grant: VerifiedGrant = Depends(grantex)):
            require_scopes(grant, "data:read")
            ...
    """
    missing = [s for s in scopes if s not in grant.scopes]
    if missing:
        raise GrantexFastAPIError(
            "SCOPE_INSUFFICIENT",
            f"Grant token is missing required scopes: {', '.join(missing)}",
            403,
        )


def grantex_exception_handler(request: Request, exc: GrantexFastAPIError) -> JSONResponse:
    """Default exception handler for GrantexFastAPIError.

    Register with::

        app.add_exception_handler(GrantexFastAPIError, grantex_exception_handler)
    """
    return JSONResponse(
        status_code=exc.status_code,
        content={"error": exc.code, "message": str(exc)},
    )

File: packages/express/src/middleware.ts
import type { Request, Response, NextFunction, RequestHandler } from 'express';
import { verifyGrantToken, GrantexTokenError } from '@grantex/sdk';
import type { VerifiedGrant } from '@grantex/sdk';
import type { GrantexMiddlewareOptions, GrantexRequest } from './types.js';
import { GrantexMiddlewareError } from './errors.js';

/**
 * Default token extractor: reads the `Authorization: Bearer <token>` header.
 */
function defaultTokenExtractor(req: Request): string | undefined {
  const header = req.headers.authorization;
  if (!header) return undefined;
  const parts = header.split(' ');
  if (parts.length !== 2 || parts[0] !== 'Bearer') return undefined;
  return parts[1];
}

function defaultErrorHandler(
  err: GrantexMiddlewareError,
  _req: Request,
  res: Response,
  _next: NextFunction,
): void {
  res.status(err.statusCode).json({
    error: err.code,
    message: err.message,
  });
}

/**
 * Creates an Express middleware that verifies Grantex grant tokens.
 *
 * On success, `req.grant` is populated with the decoded `VerifiedGrant`
 * and the next handler is called. On failure, a JSON error response is
 * sent (or your custom `onError` handler is invoked).
 *
 * @example
 * ```typescript
 * import express from 'express';
 * import { requireGrantToken } from '@grantex/express';
 *
 * const app = express();
 * const JWKS_URI = 'https://grantex-auth-dd4mtrt2gq-uc.a.run.app/.well-known/jwks.json';
 *
 * // Protect all routes under /api
 * app.use('/api', requireGrantToken({ jwksUri: JWKS_URI }));
 *
 * app.get('/api/calendar', (req, res) => {
 *   // req.grant is available here
 *   res.json({ principalId: req.grant.principalId, scopes: req.grant.scopes });
 * });
 * ```
 */
export function requireGrantToken(options: GrantexMiddlewareOptions): RequestHandler {
  const {
    jwksUri,
    tokenExtractor = defaultTokenExtractor,
    clockTolerance,
    audience,
    onError = defaultErrorHandler,
  } = options;

  return async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    const token = tokenExtractor(req);

    if (!token) {
      const err = new GrantexMiddlewareError(
        'TOKEN_MISSING',
        'No grant token found. Send a Grantex JWT in the Authorization header as "Bearer <token>".',
        401,
      );
      onError(err, req, res, next);
      return;
    }

    try {
      const grant = await verifyGrantToken(token, {
        jwksUri,
        ...(clockTolerance !== undefined ? { clockTolerance } : {}),
        ...(audience !== undefined ? { audience } : {}),
      });

      (req as GrantexRequest).grant = grant;
      next();
    } catch (err) {
      if (err instanceof GrantexTokenError) {
        const isExpired = err.message.includes('exp');
        const code = isExpired ? 'TOKEN_EXPIRED' : 'TOKEN_INVALID';
        const statusCode = 401;
        const mwErr = new GrantexMiddlewareError(code, err.message, statusCode);
        onError(mwErr, req, res, next);
        return;
      }
      next(err);
    }
  };
}

/**
 * Creates an Express middleware that checks whether the verified grant
 * contains **all** of the required scopes.
 *
 * Must be used **after** `requireGrantToken()` so that `req.grant` is set.
 *
 * @example
 * ```typescript
 * import { requireGrantToken, requireScopes } from '@grantex/express';
 *
 * app.get(
 *   '/api/calendar',
 *   requireGrantToken({ jwksUri: JWKS_URI }),
 *   requireScopes('calendar:read'),
 *   (req, res) => {
 *     res.json({ events: getCalendarEvents(req.grant.principalId) });
 *   },
 * );
 *
 * // Multiple scopes — the token must contain ALL of them
 * app.post(
 *   '/api/calendar/events',
 *   requireGrantToken({ jwksUri: JWKS_URI }),
 *   requireScopes('calendar:read', 'calendar:write'),
 *   (req, res) => {
 *     // ...
 *   },
 * );
 * ```
 */
export function requireScopes(...scopes: string[]): RequestHandler {
  return (req: Request, res: Response, next: NextFunction): void => {
    const grant = (req as GrantexRequest).grant;

    if (!grant) {
      const err = new GrantexMiddlewareError(
        'TOKEN_MISSING',
        'requireScopes() must be used after requireGrantToken(). No grant found on request.',
        500,
      );
      res.status(err.statusCode).json({ error: err.code, message: err.message });
      return;
    }

    const missing = scopes.filter((s) => !grant.scopes.includes(s));

    if (missing.length > 0) {
      const err = new GrantexMiddlewareError(
        'SCOPE_INSUFFICIENT',
        `Grant token is missing required scopes: ${missing.join(', ')}`,
        403,
      );

      res.status(err.statusCode).json({ error: err.code, message: err.message });
      return;
    }

    next();
  };
}

/**
 * Creates a Grantex middleware instance with pre-configured options.
 * Useful when you want to share the same JWKS URI and settings
 * across many routes without repeating yourself.
 *
 * @example
 * ```typescript
 * import { createGrantex } from '@grantex/express';
 *
 * const grantex = createGrantex({
 *   jwksUri: 'https://grantex-auth-dd4mtrt2gq-uc.a.run.app/.well-known/jwks.json',
 *   clockTolerance: 5,
 * });
 *
 * // Short, clean route definitions
 * app.get('/api/calendar', grantex.requireToken(), grantex.requireScopes('calendar:read'), handler);
 * app.get('/api/email',    grantex.requireToken(), grantex.requireScopes('email:read'),    handler);
 * ```
 */
export function createGrantex(options: GrantexMiddlewareOptions) {
  return {
    /**
     * Middleware that verifies the grant token.
     * Uses the options provided to `createGrantex()`.
     * You can pass additional per-route options to override.
     */
    requireToken(overrides?: Partial<GrantexMiddlewareOptions>): RequestHandler {
      return requireGrantToken({ ...options, ...overrides });
    },

    /**
     * Middleware that checks required scopes on `req.grant`.
     * Must be used after `requireToken()`.
     */
    requireScopes(...scopes: string[]): RequestHandler {
      return requireScopes(...scopes);
    },
  };
}
